ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -0,0 +1,807 @@
+        throws IOException, MessageFormatException {
+      if (!successfullyDeserialized) {
+        BlobData blobData = MessageFormatRecord.deserializeBlob(payload);
+        BlobType blobType = blobData.getBlobType();
+        chunkIndexToBuffer = new TreeMap<>();
+        if (blobType == BlobType.MetadataBlob) {
+          ByteBuffer serializedMetadataContent = blobData.getStream().getByteBuffer();
+          List<StoreKey> keys =
+              MetadataContentSerDe.deserializeMetadataContentRecord(serializedMetadataContent, blobIdFactory);
+          chunkIdIterator = keys.listIterator();
+          numChunksTotal = keys.size();
+          dataChunks = new GetChunk[Math.min(keys.size(), NonBlockingRouter.MAX_IN_MEM_CHUNKS)];
+          for (int i = 0; i < dataChunks.length; i++) {
+            dataChunks[i] = new GetChunk(chunkIdIterator.nextIndex(), (BlobId) chunkIdIterator.next());
+          }
+        } else {
+          chunkIdIterator = null;
+          numChunksTotal = 1;
+          dataChunks = null;
+          chunkIndexToBuffer.put(0, blobData.getStream().getByteBuffer());
+          numChunksRetrieved = 1;
+        }
+        successfullyDeserialized = true;
+        state = ChunkState.Complete;
+      } else {
+        // Currently, regardless of the successTarget, only the first successful response is honored. Subsequent ones
+        // are ignored. If ever in the future, we need some kind of reconciliation, this is the place
+        // to do that. (Only after the reconciliation will the state be marked as complete).
+      }
+    }
+
+    /**
+     * {@inheritDoc}
+     * <br>
+     * Receiving a {@link ServerErrorCode#Blob_Deleted}, {@link ServerErrorCode#Blob_Expired} or
+     * {@link ServerErrorCode#Blob_Not_Found} is not unexpected for the first chunk, unlike for subsequent chunks.
+     */
+    @Override
+    void processServerError(ServerErrorCode errorCode) {
+      switch (errorCode) {
+        case Blob_Deleted:
+          logger.trace("Requested blob was deleted");
+          setChunkException(new RouterException("Server returned: " + errorCode, RouterErrorCode.BlobDeleted));
+          break;
+        case Blob_Expired:
+          logger.trace("Requested blob is expired");
+          setChunkException(new RouterException("Server returned: " + errorCode, RouterErrorCode.BlobExpired));
+          break;
+        case Blob_Not_Found:
+          logger.trace("Requested blob was not found on this server");
+          setChunkException(new RouterException("Server returned: " + errorCode, RouterErrorCode.BlobDoesNotExist));
+          break;
+        default:
+          setChunkException(
+              new RouterException("Server returned: " + errorCode, RouterErrorCode.UnexpectedInternalError));
+          break;
+      }
+    }
+  }
+
+  /**
+   * Different states of a GetChunk.
+   */
+  enum ChunkState {
+    /**
+     * The GetChunk is free and can be assigned to hold a chunk of an object.
```

Comment:
will use "overall blob" like I've used elsewhere (it does not necessarily have to be a composite blob).